File: backend/src/collectors/fact-collector.ts
import * as cron from 'node-cron';
import { prisma } from '../db/client.js';
import * as dockerAdapter from '../adapters/docker.js';
import * as proxmoxAdapter from '../adapters/proxmox.js';
import {
    normalizeDockerContainer,
    normalizeDockerStats,
    normalizeProxmoxNode,
    normalizeProxmoxVM,
    NormalizedFact
} from './normalizers.js';
import { hasProxmox } from '../config/index.js';

// Collection interval in seconds (default: 30s)

// Collection interval in seconds (default: 30s)
const COLLECTION_INTERVAL_SECONDS = parseInt(process.env.FACT_COLLECTION_INTERVAL || '30', 10);

// Track if collector is running
let isRunning = false;
let cronJob: ReturnType<typeof cron.schedule> | null = null;

/**
 * Collect facts from all adapters and store in database
 */
export async function collectFacts(): Promise<{ docker: number; proxmox: number }> {
    const results = { docker: 0, proxmox: 0 };
    const factsToStore: NormalizedFact[] = [];

    // Collect Docker facts
    try {
        if (await dockerAdapter.isDockerAvailable()) {
            const containers = await dockerAdapter.listContainers();

            for (const container of containers) {
                // Get detailed info with restart count
                try {
                    const info = await dockerAdapter.getContainerInfo(container.id);
                    factsToStore.push(normalizeDockerContainer(info));

                    // Only get stats for running containers
                    if (info.state === 'running') {
                        try {
                            const stats = await dockerAdapter.getContainerStats(container.id);
                            factsToStore.push(normalizeDockerStats(container.id, stats));
                        } catch (statsError) {
                            // Stats may fail for some containers, continue
                        }
                    }
                    results.docker++;
                } catch (infoError) {
                    // Use basic container info if inspect fails
                    factsToStore.push(normalizeDockerContainer(container));
                    results.docker++;
                }
            }
            console.log(`📦 Collected ${results.docker} Docker container facts`);
        }
    } catch (error) {
        console.error('Fact collector: Docker collection failed', error);
    }

    // Collect Proxmox facts
    try {
        if (hasProxmox() && await proxmoxAdapter.isProxmoxAvailable()) {
            const { nodes, vms } = await proxmoxAdapter.listAllResources();

            // Store node facts
            for (const node of nodes) {
                factsToStore.push(normalizeProxmoxNode(node));
                results.proxmox++;
            }

            // Store VM/LXC facts
            for (const vm of vms) {
                factsToStore.push(normalizeProxmoxVM(vm));
                results.proxmox++;
            }
            console.log(`🖥️ Collected ${results.proxmox} Proxmox facts (${nodes.length} nodes, ${vms.length} VMs/LXCs)`);
        }
    } catch (error) {
        console.error('Fact collector: Proxmox collection failed', error);
    }

    // Store all facts in database
    if (factsToStore.length > 0) {
        try {
            await prisma.fact.createMany({
                data: factsToStore.map((fact) => ({
                    resourceRef: fact.resourceRef,
                    factType: fact.factType,
                    value: fact.value,
                    source: fact.source,
                })),
            });
            console.log(`💾 Stored ${factsToStore.length} facts in database`);
        } catch (error) {
            console.error('Fact collector: Failed to store facts', error);
        }
    }

    return results;
}

/**
 * Start the fact collector scheduler
 */
export function startCollector(): void {
    if (isRunning) {
        console.log('Fact collector already running');
        return;
    }

    // Run immediately on startup
    console.log(`🚀 Starting fact collector (interval: ${COLLECTION_INTERVAL_SECONDS}s)`);
    collectFacts().catch(console.error);

    // Schedule recurring collection
    const cronExpression = `*/${COLLECTION_INTERVAL_SECONDS} * * * * *`;
    cronJob = cron.schedule(cronExpression, () => {
        collectFacts().catch(console.error);
    });

    isRunning = true;
}

/**
 * Stop the fact collector scheduler
 */
export function stopCollector(): void {
    if (cronJob) {
        cronJob.stop();
        cronJob = null;
    }
    isRunning = false;
    console.log('🛑 Fact collector stopped');
}

/**
 * Get collector status
 */
export function getCollectorStatus(): { running: boolean; intervalSeconds: number } {
    return {
        running: isRunning,
        intervalSeconds: COLLECTION_INTERVAL_SECONDS,
    };
}

File: backend/app/collectors/fact_collector.py
"""Fact Collector - collects normalized observations from adapters."""

from datetime import datetime, timedelta
from typing import Any
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.storage.models import Fact
from app.adapters import docker_adapter, proxmox_adapter


class FactCollector:
    """Collects and stores normalized facts from infrastructure adapters."""
    
    async def collect_all(self, db: AsyncSession) -> dict[str, int]:
        """Collect facts from all adapters."""
        counts = {
            "docker": 0,
            "proxmox": 0,
        }
        
        counts["docker"] = await self.collect_docker_facts(db)
        counts["proxmox"] = await self.collect_proxmox_facts(db)
        
        return counts
    
    async def collect_docker_facts(self, db: AsyncSession) -> int:
        """Collect facts from Docker containers."""
        if not docker_adapter.is_connected:
            return 0
        
        containers = await docker_adapter.list_containers(all=True)
        count = 0
        
        for container in containers:
            # Container status fact (Observation only)
            await self._store_fact(
                db,
                resource_ref=container["resource_ref"],
                fact_type="container_status",
                value={
                    "status": container["status"],
                    "restart_count": container["restart_count"],
                    "health": container.get("health"),
                    "image": container["image"],
                    "name": container["name"],
                },
                source="docker",
            )
            count += 1
        
        return count
    
    async def collect_proxmox_facts(self, db: AsyncSession) -> int:
        """Collect facts from Proxmox nodes, VMs, and LXCs."""
        if not proxmox_adapter.is_connected:
            return 0
        
        count = 0
        
        # Collect node facts
        nodes = await proxmox_adapter.list_nodes()
        for node in nodes:
            await self._store_fact(
                db,
                resource_ref=node["resource_ref"],
                fact_type="node_status",
                value={
                    "status": node["status"],
                    "cpu": node.get("cpu"),
                    "mem": node.get("mem"),
                    "maxmem": node.get("maxmem"),
                    "uptime": node.get("uptime"),
                },
                source="proxmox",
            )
            count += 1
        
        # Collect VM facts
        vms = await proxmox_adapter.list_vms()
        for vm in vms:
            await self._store_fact(
                db,
                resource_ref=vm["resource_ref"],
                fact_type="vm_status",
                value={
                    "status": vm["status"],
                    "name": vm["name"],
                    "node": vm["node"],
                    "cpu": vm.get("cpu"),
                    "mem": vm.get("mem"),
                    "uptime": vm.get("uptime"),
                },
                source="proxmox",
            )
            count += 1
        
        # Collect LXC facts
        lxcs = await proxmox_adapter.list_lxcs()
        for lxc in lxcs:
            await self._store_fact(
                db,
                resource_ref=lxc["resource_ref"],
                fact_type="lxc_status",
                value={
                    "status": lxc["status"],
                    "name": lxc["name"],
                    "node": lxc["node"],
                    "cpu": lxc.get("cpu"),
                    "mem": lxc.get("mem"),
                    "uptime": lxc.get("uptime"),
                },
                source="proxmox",
            )
            count += 1
        
        return count
    
    async def get_recent_facts(
        self, 
        db: AsyncSession, 
        resource_ref: str | None = None,
        fact_type: str | None = None,
        hours: int = 24,
    ) -> list[Fact]:
        """Get recent facts, optionally filtered."""
        since = datetime.utcnow() - timedelta(hours=hours)
        
        query = select(Fact).where(Fact.timestamp >= since)
        
        if resource_ref:
            query = query.where(Fact.resource_ref == resource_ref)
        if fact_type:
            query = query.where(Fact.fact_type == fact_type)
        
        query = query.order_by(Fact.timestamp.desc()).limit(100)
        
        result = await db.execute(query)
        return list(result.scalars().all())
    
    async def _store_fact(
        self,
        db: AsyncSession,
        resource_ref: str,
        fact_type: str,
        value: dict[str, Any],
        source: str,
    ) -> Fact:
        """Store a new fact in the database."""
        fact = Fact(
            resource_ref=resource_ref,
            fact_type=fact_type,
            value=value,
            source=source,
            timestamp=datetime.utcnow(),
        )
        db.add(fact)
        return fact


# Singleton
fact_collector = FactCollector()
